File: linen-starter/src/main/java/x7c1/linen/SourceRowAdapter.java
package x7c1.linen;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import java.util.ArrayList;
import java.util.List;

public class SourceRowAdapter extends RecyclerView.Adapter<SourceRowHolder>{

	private final LayoutInflater inflater;
	private final List<Source> sources;

	public SourceRowAdapter(Context context) {
		this.sources = createDummyList();
		this.inflater = LayoutInflater.from(context);
	}

	private List<Source> createDummyList(){
		List<Source> list = new ArrayList<>();
		for (int i = 0; i < 100; i++){
			list.add(new Source("sample-title-" + i, "sample-description-" + i));
		}
		return list;
	}

	@Override
	public SourceRowHolder onCreateViewHolder(ViewGroup parent, int viewType) {
		return new SourceRowHolder(inflater.inflate(R.layout.source_row, parent, false));
	}

	@Override
	public void onBindViewHolder(SourceRowHolder holder, int position) {
		Source source = sources.get(position);
		holder.title.setText(source.getTitle());
		holder.description.setText(source.getDescription());
	}

	@Override
	public int getItemCount() {
		return sources.size();
	}


}

File: linen-modern/src/main/scala/x7c1/linen/modern/display/SourceRowAdapter.scala
package x7c1.linen.modern.display

import android.support.v7.widget.RecyclerView.Adapter
import android.view.ViewGroup
import x7c1.linen.glue.res.layout.SourceRow
import x7c1.linen.modern.accessor.UnreadSourceAccessor
import x7c1.linen.modern.struct.UnreadSource
import x7c1.wheat.ancient.resource.ViewHolderProvider
import x7c1.wheat.modern.decorator.Imports._


class SourceRowAdapter(
  sourceAccessor: UnreadSourceAccessor,
  sourceSelectedListener: OnSourceSelectedListener,
  viewHolderProvider: ViewHolderProvider[SourceRow]) extends Adapter[SourceRow]{

  override def onCreateViewHolder(parent: ViewGroup, viewType: Int) = {
    viewHolderProvider inflateOn parent
  }

  override def onBindViewHolder(holder: SourceRow, position: Int) = {
    sourceAccessor.findAt(position) foreach { source =>
      holder.title.text = source.title
      holder.description.text = source.description
      holder.itemView onClick { view =>
        val event = SourceSelectedEvent(position, source)
        sourceSelectedListener onSourceSelected event
      }
    }
  }

  override def getItemCount = sourceAccessor.length
}

trait OnSourceSelectedListener {
  def onSourceSelected(event: SourceSelectedEvent): Unit
}

case class SourceSelectedEvent (position: Int, source: UnreadSource){
  def dump: String = s"sourceId:${source.id}, position:$position"
}
